File: Practicas/Qytetet/QytetetRuby/lib/especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module ModeloQytetet
  class Jugador
  end
  
  class Especulador < Jugador
    attr_accessor :fianza

    def self.copia(j, f)
      especulador = super(j)
      especulador.fianza = f
      especulador
    end

    
    def pagar_impuesto
      coste_impuesto =  @casilla_actual.coste / 2
      modificar_saldo(-coste_impuesto)
    end

    def debo_ir_a_carcel
      debo = super && !pagar_fianza
      debo
    end

    def convertirme(fianza)
      self
    end

    def pagar_fianza
      tengo_s = tengo_saldo(@fianza)
      if(tengo_s)
        modificar_saldo(-@fianza)
        @encarcelado = false
      end
      tengo_s
    end

    def puedo_edificar_casa(titulo)
      num_casas = titulo.num_casas
      ((num_casas < 8) && es_de_mi_propiedad(titulo))
    end

    def puedo_edificar_hotel(titulo)
      num_casas = titulo.num_casas
      num_hoteles = titulo.num_hoteles
      (num_casas >=4 && num_hoteles < 8 && es_de_mi_propiedad(titulo))
    end  

    def to_s
      "ESPECULADOR (#{@fianza}) -> " + super.to_s
    end

  end
end

File: Practicas/Qytetet/QytetetJava/src/modeloqytetet/Especulador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author enrique
 */
public class Especulador extends Jugador{
    private int fianza;
    
    Especulador(Jugador jugador, int fianza)
    {
        super(jugador);
        this.fianza = fianza;
    }
    
    @Override
    protected void pagarImpuesto()
    {
        int costeImpuesto = this.getCasillaActual().getCoste() / 2; 
        modificarSaldo(-costeImpuesto);
    }
    
    @Override
    protected boolean deboIrACarcel()
    {
        boolean debo = super.deboIrACarcel() && !pagarFianza();
        return debo;
    }
    
    @Override
    protected Especulador convertirme(int fianza)
    {
        return this;
    }
    
    private boolean pagarFianza()
    {
        boolean tengoSaldo = tengoSaldo(fianza);
        if(tengoSaldo)
        {
            modificarSaldo(-fianza);
            setEncarcelado(false);
        }
        return tengoSaldo;
    }
    
    @Override
    protected int getFactorEspeculador()
    {
        return fianza;
    }
    
    @Override
    protected boolean puedoEdificarCasa(TituloPropiedad titulo)
    {
        int numCasas = titulo.getNumCasas();
        // Hemos cambiado la visibilidad de esDeMiPropiedad de private a protected
        return ((numCasas < 8) && esDeMiPropiedad(titulo));
    }
    
    @Override
    protected boolean puedoEdificarHotel(TituloPropiedad titulo)
    {
        int numCasas = titulo.getNumCasas();
        int numHoteles = titulo.getNumHoteles();
        // Hemos cambiado la visibilidad de esDeMiPropiedad de private a protected
        return (numCasas >=4 && numHoteles < 8 && esDeMiPropiedad(titulo));
    }
    
    @Override
    public String toString() {
        return "ESPECULADOR (" + fianza + ") -> " + super.toString();
    }
}
